spring-xsuaa/src/main/java/com/sap/cloud/security/xsuaa/autoconfiguration/XsuaaAutoConfiguration.java

```diff
@@ -39,4 +44,45 @@ public XsuaaServiceConfiguration xsuaaServiceConfiguration() {
 			return new XsuaaServiceConfigurationDefault();
 		}
 	}
+
+	/**
+	 * Creates a new {@link XsuaaTokenFlows} bean that applications can auto-wire
+	 * into their controllers to perform a programmatic token flow exchange.
+	 * 
+	 * @param restTemplate
+	 *            - the {@link RestTemplate} to use for the token flow exchange.
+	 * @param decoder
+	 *            - the decoder used for the tokens retrieved via the token flows.
+	 * @return the {@link XsuaaTokenFlows} API.
+	 */
+	@Bean
+	@ConditionalOnMissingBean
+	public XsuaaTokenFlows xsuaaTokenFlows(RestTemplate restTemplate, VariableKeySetUriTokenDecoder decoder) {
+		return new XsuaaTokenFlows(restTemplate, decoder);
+	}
+
+	/**
+	 * Creates a {@link VariableKeySetUriTokenDecoder} instance based on a
+	 * {@code NimbusJwtDecoderJwkSupport} implementation which is used by the
+	 * {@link XsuaaTokenFlows} bean.
+	 * 
+	 * @return the {@link VariableKeySetUriTokenDecoder} instance.
+	 */
+	@Bean
+	@ConditionalOnMissingBean
+	public VariableKeySetUriTokenDecoder xsuaaTokenDecoder() {
```

Comment:
The answer is given in class `VariableKeySetUriTokenDecoder`:

```
/**
 * Token decoder interface to provide for 
 * dependency injection of actual token decoder
 * implementation. 
 * 
 * This interface is used to implement a token decoder
 * that can decode tokens from various sources.
 * Since decoding requires the public keys of the token's origin
 * the {@code keySetUri} needs to be specified before decoding.
 * 
 * <b>Note:</b> the standard JwtDecoder exposed as a bean in class 
 * {@link XsuaaResourceServerJwkAutoConfiguration} is referring to 
 * a single Key Set URI, only. This interface defines an API for a
 * JwtDecoder which can be given changing key set URIs.
 */
```
The `XsuaaResourceServerJwkAutoConfiguration.java` uses a keyset URI which (may be tenant-specific but) always points to XSUAA. The Token Flows API can potentially be used with another keyset endpoint / OAuth server (which is why the Token Flows all have two constructors, one of which is taking the actual endpoints to communicate with). Therefore, we need another TokenDecoder here, and the name prefix `VariableKeySetUri...` tried to express that.